src/coreclr/jit/optimizer.cpp

```diff
@@ -6961,8 +6989,8 @@ bool Compiler::optHoistThisLoop(unsigned lnum, LoopHoistContext* hoistCtxt, Basi
             JITDUMP("  -- odd, we didn't reach entry from exit via dominators. Only considering hoisting in entry "
                     "block " FMT_BB "\n",
                     pLoopDsc->lpEntry->bbNum);
-            defExec.Reset();
-            preHeadersList = existingPreHeaders;
+            defExec.Pop(defExec.Height() - childLoopPreHeaders);
+            assert(defExec.Height() == childLoopPreHeaders);
```

Comment:
fwiw, this `if` appears to be dead code. It's never hit in SPMI anyway. And that makes sense: walking up the immediate dominators from the single loop exit block (or any loop exit block, for that matter, though we don't track non-single-exit blocks) should always reach the loop entry block. Or, in other words, the (single) loop entry should dominate the (single) loop exit.

Reply:
Is it worth adding `assert(false)` here and run SPMI for all configurations and get rid of it if we don't hit it?

Reply:
I ran an experiment and we actually do hit this case in a very few x86 tests. We should probably augment our loop recognition to reject those loops, in which the "exit" block of the loop is an EH handler. Presumably if the loop has a normal exit as well as an EH exit, on x86 we won't get here because we only get here for "single exit" loops. So it requires an "infinite" loop with the only exit being from the EH handler.

Opened https://github.com/dotnet/runtime/issues/84222 to track.